[refactor] 예외 처리 재설계 & 데이터 옵저빙 기술 스택 변경

yrsel opened the pull request:
## Issues
- closed #507

## ✔️  Check-list
- [x] : Label을 지정해 주세요.
- [x] : Merge할 브랜치를 확인해 주세요.

## 🗒️ Work Description
- 튜립 Result 기존 설계 수정
- 서버에서 내려준 tag 기반으로 사용자에게 문제 원인을 전달할 수 있는 구조로 수정

---

## 문제 상황 

기존 TuripCustomResult가 data 패키지에 위치해 있지만, presentation 에서도 사용하고 있는 구조였고 서버에서 내려주는 에러코드, 메시지를 사용하기 어려운 구조였습니다.
앞으로 진행할 모듈화 과정에서 적절한 관심사 분리가 이뤄져야 원활한 작업이 진행 될 것 같아 재구조화 했습니다.

## 해결 방법

재구성하면서 화면에 대한 로딩, 에러, 정상, 데이터 없을 때에 대한 처리가 미흡하거나 없는 부분 모두 적용했습니다.
이 과정에서 뷰모델에서 기존 분산된 필드들을 상황에 맞게 모으며 Flow로 전환하였습니다.( UiState(상태), UiEffect(행동에 대한 이벤트 처리))

네트워크, 서버 에러 발생 시 
전체 화면을 변경해야하는 경우에는 uiState(StateFlow) 값에 따라 화면에 대한 상태를 분기하였고, 
다이얼로그 또는 이벤트 처리가 적합(ex, 컨텐츠 상세 화면에서 컨텐츠 찜을 누르는 경우)한 경우 uiEffect(Channel)로 스낵바를 통해 정보를 제공하도록 하였습니다.

현재 재구조화한 구조입니다.
제리의 모듈화 설계 과정에서 필요하다고 느낀 core-result 모듈은 일단 패키지로 분리했습니다.
```
core/result : ErrorType, TuripResult, TuripResultExtensions
data/result : ErrorResponse, ErrorExtensions, SafeApiCall
ui/common/error : UiError, UiErrorExtensions, ErrorUiState, ErrorUiModel

core/result
ErrorType : 도메인 에러 타입
TuripResult : 튜립 서버와 클라이언트간 에러를 처리하기 위한 커스텀 클래스
TuripResultExtensions : TuripCustomResult 를 활용하기 위한 확장함수들 

data/result
ErrorResponse : 서버에서 내려주는 에러 정보 
ErrorExtensions : 서버에서 내려주는 에러 정보를 도메인(코어) 객체로 변환 
SafeApiCall : Remote API TuripResult로 변환하는 함수들

ui/common/error
UiError : 사용자에게 에러 정보를 전달할 필요가 있는 에러들의 모음
UiErrorExtensions : 도메인 에러 타입을 UiError 타입으로 변환
ErrorUiState : 네트워크, 서버 에러에 대한 상태 표현하는 모델
ErrorUiModel :  네트워크, 서버 에러에 대한 에러 정보들을 담고 있는 모델
```

- TuripCustomResult 라는 이름을 가지고 있는데, Custom 이라는 이름이 불필요하게 느껴저서 제거하는 건 어떤가 의견 드립니다.  
-> 모두 동의하셔서 반영했습니다! 😊 
[패키지 이동 및 TuripResult로 변경](https://github.com/woowacourse-teams/2025-Turip/pull/508/commits/59693d4a9f9793d655494f645c2a0c7a6393d5df)

- 로깅은 api 호출하면 catch 할 때 처리하는 게 좋을 것 같아서, 기존 TuripCustomResult 내부에서 제거하고 아직은 로깅은 추가하지 않았습니다.  SafeApiCall 파일의 catch 블럭에서 하는건 어떤가 의견드려요. ( + 로거에 대한 수정 작업도 필요해보이는데 어떤지 궁금합니다. )  
-> 로깅, 로거에 대한 작업은 따로 이슈 파서 작업하면 좋을 것 같네요 !  

+ 로딩을 거쳐서 API를 호출하는 함수들이 응답을 빠른 시간에 받다보니 state가 순식간에 변경되며 깜빡이는 현상이 있습니다. 어떤 방법으로 해결해야 좋을 지 의견 부탁드립니다..!  
-> 하나의 방법으로는 loading 상태 변경을 async 코루틴으로 감싸고 내부에서 delay(300) 정도로 설정하고 
성공 또는 실패할 때 해당 loading Job을 cancel 하는 방식이 있는 것 같아요!(DI를 통한 LoadingPolicy 주입 또는 확장함수를 통해 해당 job 성공,취소에 대한 로직을 제공할 수도 있을 것 같아요) 
이 방법은 모든 ViewModel 내부에서 서버 API 호출하는 곳마다 감싸서 사용해야 되는 문제가 발생합니다.. 다른 방법은 어떤 방법들이 있을 지 의견 부탁드립니다 🙇‍♂️
-> when 문 분기하는 방식을 수정해서 해결했습니다!

## 📷 Screenshot

## 📚 Reference


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **새로운 기능**
  * 화면 단위 uiState/uiEffect 패턴 도입으로 일회성 효과(네비게이션·스낵바)와 상태(로딩/빈/성공/오류) 일관화

* **개선사항**
  * 오류 UI 통합(에러뷰·스낵바) 및 재시도 흐름 표준화
  * 네트워크·서버·인증 오류 분류 도입으로 메시지와 처리가 더 정교해짐
  * 로딩/빈 상태 표현 강화 및 용어 통일(Travel → Trip)로 UI 일관성 향상

* **버그 수정**
  * 토큰 만료 처리와 로그인 전환 안정성 개선

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

yrsel committed:
refactor: safeApiCall 재설계

- 상태코드 204 인 경우, 모든 api가 Unit 반환하고 있기에 body가 비어있는 경우 존재하지 않아 코드 수정
- response 호출 실패한 경우 HttpError 반환
- 코루틴 예외는 구조화된 동시성을 깨뜨리지 않도록 그대로 예외 전달
- 네트워크 에러인 경우 NetworkError 반환 (이 부분은 시스템 네트워크 상태 관찰하고 판단하는 구조로 추가 보완 필요)
- 그 외 예외인 경우 UnknownError 반환

yrsel committed:
feat: 튜립 에러 타입 정의

yrsel committed:
refactor: 튜립 Result 객체 리팩터링

- 로깅은 Result 예외 호출 시점에 하도록 책임 분리
- 튜립 Result 확장함수 별도 파일로 분리

yrsel committed:
feat: 서버 에러 활용할 수 있도록 Extensions 추가

yrsel committed:
refactor: 튜립 Result 호출 진입점 추가 정의 및 기존 함수 수정

- ErrorResponse에서 도메인에러 타입으로 변환하는 부분 함수로 분리
- 커스터 에러 외 다른 예외의 경우 Unknown으로 처리되어서 판단할 수 있도록 Throwable 전달 받도록 수정
- Local API 호출 시에서 튜립 Result 사용할 수 있도록 함수 정의

yrsel commented on android/app/src/main/java/com/on/turip/data/result/SafeApiCall.kt:
```diff
@@ -1,31 +1,48 @@
-                }
-            }
-        } else {
-            TuripCustomResult.HttpError(response.code())
+            @Suppress("UNCHECKED_CAST")
+            return TuripCustomResult.Success(response.body() as T)
+        }
```
현재 설계 상 204일 때도 Unit 값을 받도록 되어있는데 이래도 문제가 발생하나요 ?
만약에 NPE 발생한다면, 개발자가 우리의 규칙을 지키지 않아서 발생한 문제니까 catch에서 로깅을 해서 사전에 확인할 수 있지 않을까요 ?

jerry8282 replied:
혹시 어디에 204일 때 Unit값을 받도록 되어 있나유,,? 찾아도 어디에 있는지 안보여서🥲

yrsel replied:
> 혹시 어디에 204일 때 Unit값을 받도록 되어 있나유,,? 찾아도 어디에 있는지 안보여서🥲

@jerry8282 
저희  성공 응답 204 반환 API 반환값이 모두 Response\<Unit> 으로 되어 있어요! 
로그아웃, 토큰 유효성 검사, 회원 데이터 마이그레이션 등등 ..

yrsel committed:
feat: Ui Error 타입 정의

yrsel committed:
refactor: CustomErrorView View binding 적용 및 리팩터링

- ViewBinding 으로 View를 탐색하도록 수정
- 하드코딩 주입 방식은 사용되지 않아 제거
- 전체 화면을 대체하는 에러는 Network, Server 에러정도 일 것으로 판단 후 처리, 다른 에러들의 경우 토스트 메시지나 다른 방식으로 사용자에게 전달하는 게 적절할 것 같습니다.

yrsel committed:
refactor: 공통 Ui Event 추가 및 수정

- CommonEvent -> CommonUiEffect 이름 수정
- 트리거 되는 조건이 아닌 행동을 표현 하도록 수정 (TokenExpiration -> NavigateToLogin)
- ShowErrorView 클래스 : 커스텀 에러뷰
- NotifyError 클래스 : 다이얼로그, 스낵바로 에러 전달

yrsel committed:
refactor: 공통 Ui Event 추가 및 수정

- CommonEvent -> CommonUiEffect 이름 수정
- 트리거 되는 조건이 아닌 행동을 표현 하도록 수정 (TokenExpiration -> NavigateToLogin)

yrsel commented on android/app/src/main/java/com/on/turip/ui/common/event/CommonUiEffect.kt:
```diff
@@ -0,0 +1,24 @@
+     */
+    data class ShowErrorView(
+        val error: UiError,
+    ) : CommonUiEffect
```
여기는 이벤트가 아니라 상태 기반으로 처리해야 해서 제거 되었습니다.

yrsel committed:
refactor: 전역에서 사용되는 UiError와 화면별 사용되는 UiError 계층 분리

yrsel committed:
refactor: ErrorHandlingContainer 제거

- ErrorHandlingContainer를 모든 Screen에서 감싸서 에러 화면을 처리하게 되면 Container에 종속적이게 되고 오히려 화면마다 에러를 달리 가져갈 때 걸림돌이 될 것 같아서 제거

yrsel committed:
feat: Server Error, Network Error에 대한 UiModel, UiState 추가

- Server Error와 Network Error 는 동시에 발생할 수 없기에 하나의 State로 관리하도록 설계

yrsel committed:
refactor: ErrorScreen uiState, uiModel 활용하도록 수정

yrsel committed:
refactor: SettingViewModel 함수명 수정

- 매개변수에 따라 변경되는 함수명이 show가 아닌 onXXX 이벤트로 느껴지도록 수정

yrsel committed:
refactor: Setting 에러 처리 로직 수정

- SettingUiEvent -> SettingUiEffect 이름 수정
- 로그아웃, 탈퇴 시 모두 로그인으로 이동하는 행동으로 표현
- 이전 에러 타입에서 새로 정의한 UiError 타입으로 판단하도록 수정

yrsel committed:
refactor: HomeScreen 새로운 에러 타입으로 리팩터링

yrsel committed:
feat: Flow 생명주기 확장함수

yrsel committed:
refactor: CustomErrorView ViewSystem에서도 Compose와 동일하게 uiState에 매핑된 리소스 사용하도록 수정

yrsel committed:
refactor: FavoriteContentFragment 새로운 에러 타입으로 리팩터링

yrsel committed:
refactor: 공유하기 기능 상태가 아닌 이벤트로 처리하도록 수정

jerry8282 commented on android/app/src/main/java/com/on/turip/data/result/ErrorExtensions.kt:
```diff
@@ -0,0 +1,39 @@
+ */
+fun ErrorResponse?.toErrorType(): ErrorType {
+    if (this == null) return ErrorType.Unknown
+    return this.tag.trim().toErrorType()
```
현재는 tag만 사용하고 있고 message는 별도로 사용하지 않고 계신 것 같네유
제가 이전에 message를 함께 받아와야겠다고 생각했던 이유는, 서버에서 의도적으로 내려주는 예외 상황에 맞는 메시지를 클라이언트에서 그대로 표현하기 위함이었습니다.

ErrorType의 body를 data class로 정의하고 그 내부에 message와 같은 필드를 두는 방식은 어떠신가용?
이렇게 하면 에러가 발생했을 때 서버에서 내려준 메시지를 그대로 활용해 출력할 수 있을 거 같아요😊

yrsel replied:
제가 생각했을 때 서버에서 내려주는 메시지는 로깅에 사용할 수 있을 거라고 생각해서 message를 포함하는 형태로 ErrorResponse를 받도록 했어요!

이 메시지를 사용자에게 보여줄 것인지 고민해봤는데 에러지만 메시지를 전송하지 않는 경우도 존재하고, 사용자에게 전달할 메시지를 가공하고 싶을 경우 UI단에서 수정이 이뤄질 거 같은데 
서버에서 내려준 메시지가 존재한다면 가공한 메시지와 서버의 메시지 둘 중 어떤 것을 써야될 지 판단하기 애매할 것 같다고 생각이 들었어요!

jerry8282 commented on android/app/src/main/java/com/on/turip/data/common/ErrorType.kt:
```diff
@@ -0,0 +1,109 @@
+
+/**
+ * domain layer
+ */
```
이게 만약 domain layer라면 이후에 TuripCustomResult가 core모듈로 가게 될텐데 그렇게 된다면
core에서 domain레이어에 있는 ErrorType을 알게 되는 거 같아요.🥲 
ErrorType을 core로 옮기는 건 어떻게 생각하시나요?

yrsel replied:
반영했습니다 ! 
반영 커밋 : https://github.com/woowacourse-teams/2025-Turip/pull/508/commits/59693d4a9f9793d655494f645c2a0c7a6393d5df

yrsel committed:
feat: 이벤트로 예외 정보 전달하는 ErrorUiEffect 구현

- 화면 전체를 에러뷰로 대체하는 것이 아닌 상황에 에러 정보를 전달하기 위한 이벤트

yrsel committed:
refactor: FavoritePlaceFolderCatalogFragment 새로운 에러 타입으로 리팩터링

jerry8282 commented on android/app/src/main/java/com/on/turip/data/common/SafeApiCall.kt:
```diff
@@ -1,31 +1,48 @@
 
-@Suppress("UNCHECKED_CAST")
-suspend inline fun <T> safeApiCall(apiCall: suspend () -> Response<T>): TuripCustomResult<T> =
+suspend fun <T> safeApiCall(apiCall: suspend () -> Response<T>): TuripCustomResult<T> {
```
```suggestion
suspend inline fun <T> safeApiCall(apiCall: suspend () -> Response<T>): TuripCustomResult<T> {
```
고차함수라 다음과 같이 inline을 붙이는 게 더 좋을 거 같아요!

yrsel replied:
고차함수긴 한데 suspend 함수고 해당 함수는 인라이닝이 성능상 이점이 없다고 하네요

inline 키워드를 붙이면 apiCall 부분의 suspend 에 
`Redundant 'suspend' modifier: lambda parameters of suspend function type uses existing continuation.`
문구가 나오고 제안하는 행동을 따라가면  apiCall 부분의 suspend를 제거하도록 되고 safeApiCall의 suspend도 사용되지 않는 상태로 표시되어서 apiCall에서는 Continuation 객체를 사용할 수 없지 않나요..?

최종적으로 inline 키워드를 붙이면 
```kotlin
inline fun <T> safeApiCall(apiCall: () -> Response<T>): TuripResult<T> {
```
이 형태로 사용하라는 것 같아요

yrsel committed:
refactor: FavoritePlaceFragment 공유하기 상태가 아닌 이벤트로 처리하도록 수정

jerry8282 commented on android/app/src/main/java/com/on/turip/data/common/SafeApiCall.kt:
```diff
@@ -1,31 +1,48 @@
+suspend fun <T> safeLocalApiCall(
+    errorType: ErrorType = ErrorType.Local.Unknown,
+    apiCall: suspend () -> T,
+): TuripCustomResult<T> {
```
아직 사용하는 곳이 없어서 잘 모르겠는데 errorType을 매개변수로 받는 이유가 무엇인가요?

yrsel replied:
로컬 예외로 사용자에게 전달할 정보가 없을 것 같아서 제거하였습니다 : )

jerry8282 reviewed (commented):
고생하셨습니다 채넛!! 궁금한 사항 몇가지 남겨보았습니다 😊

yrsel committed:
refactor: FavoritePlace xml 설정 오류 수정

yrsel committed:
refactor: FavoritePlaceFragment 새로운 에러 타입으로 리팩터링

yrsel committed:
refactor: 로컬에 대한 ErrorType, api함수 제거

- 로컬에러로 사용자에게 전달할 정보가 없다고 판단하고 제거
- 로컬 API는 코틀린 Result 사용

yrsel committed:
refactor: SearchHistory data layer 리팩터링

- DataSource -> Repository Result 객체로 반환만하고 사용이 안되고 있어서 DataSource에서는 원시타입 반환, Repository에서 로깅 후 반환하도록 수정

yrsel committed:
refactor: SearchActivity xml 사용되지 않는 로딩 TextView 제거

yrsel committed:
feat: 검색화면 상태에 따른 조건 분기를 위해 UiState 정의

- 필드로 에러, 로딩, 정상, 비어있는 상태를 관리하다보니 UI 관리 복잡도, 상태 중복 처리 등의 문제로 하나의 상태로 정의

yrsel committed:
refactor: SearchViewModel 검색 정보 불러오기 리팩터링

- 병렬로 실행 및 상태 관리할 수 있는 구조로 변경

yrsel committed:
refactor: SearchActivity 새로운 에러 타입으로 리팩터링

yrsel committed:
refactor: RegionResultActivity 새로운 에러타입으로 리팩터링

yrsel committed:
refactor: Travel -> Trip 수정 안된 부분 반영

- Travel 이라는 과거 단어를 Trip 으로 모두 대체

m6z1 commented on android/app/src/main/java/com/on/turip/data/result/ErrorExtensions.kt:
```diff
@@ -0,0 +1,39 @@
+ */
+fun ErrorResponse?.toErrorType(): ErrorType {
+    if (this == null) return ErrorType.Unknown
+    return this.tag.trim().toErrorType()
```
kotlin 스럽게 this 지우는 거 어때여 zzz

yrsel replied:
반영 커밋 : https://github.com/woowacourse-teams/2025-Turip/pull/508/commits/ae5009ea92596baf7934482d185a3b598efa480e

m6z1 commented on android/app/src/main/java/com/on/turip/data/result/ErrorExtensions.kt:
```diff
@@ -0,0 +1,39 @@
+/**
+ * 서버 에러 tag 매핑
+ */
+private fun String.toErrorType(): ErrorType =
```
와우 노가다 힘드셨을 것 같아요 고생하셨어요 ㅠ

m6z1 commented on android/app/src/main/java/com/on/turip/core/result/ErrorType.kt:
노가다 버전 2......

m6z1 commented on android/app/src/main/java/com/on/turip/data/common/TuripResultExtensions.kt:
```diff
@@ -0,0 +1,75 @@
+
+        is TuripCustomResult.Failure -> {
+            this
+        }
```
이런 건 
```suggestion
        is TuripCustomResult.Failure -> this
```

이렇게 써도 되지 않나염 ?

yrsel replied:
반영 커밋 : https://github.com/woowacourse-teams/2025-Turip/pull/508/commits/ae5009ea92596baf7934482d185a3b598efa480e

m6z1 commented on android/app/src/main/java/com/on/turip/core/result/TuripResult.kt:
custom 빼는 거 좋습니다!

m6z1 commented on android/app/src/main/java/com/on/turip/data/common/TuripResultExtensions.kt:
```diff
@@ -0,0 +1,75 @@
+inline fun <T> TuripCustomResult<T>.onFailureWithCause(action: (errorType: ErrorType, cause: Throwable?) -> Unit): TuripCustomResult<T> {
+    if (this is TuripCustomResult.Failure) action(errorType, cause)
+    return this
+}
```
로깅용과 없는 경우를 분리한 이유가 궁금해요! 어떤 이유에서 cause 를 optional 하게 두었는지 ? 채넛의 의도가 궁금합니닷

yrsel replied:
로깅용과 코드가 섞이지 않게 관리해보려고 분리했는데 로깅을 어디 위치에서 어떤 식으로 하는 게 좋을 지 고민이에요!
혹시나 예외를 던지지 않을 상황도 있으려나 싶어서 optional 선택지를 줬는데 굳이 일까요?

m6z1 replied:
예외를 던지지 않을 상황이 있다는 게 예시로 들면 어떤 상황일까요 ??
제가 생각했을 땐 당장 사용하고 있지 않기도 하고, 예외를 던지지 않을 상황이 잘 와닿지 않아서 예시 들어주시면 다시 생각해보겠습니다!

yrsel replied:
로깅에 예외타입만 사용하고 싶을 수도 있어서 예외에 대한 스택트레이스는 사용되지 않을 수 있지 않을까 생각했었습니다!
그렇다고 해도 사용하는 쪽에서 활용안하는 방식이 더 적절한 거 같네요 😅

optional 하지 않게 수정하였습니다!
반영 커밋 : cc83dbc0cd0bc7eab2177dd264f63dbe64b5ccb5

m6z1 commented on android/app/src/main/java/com/on/turip/data/common/UiError.kt:
이름이 UiError 인데 data 레이어 자체에서 ui 라는 이름을 가지고 있는 게 어색한 것 같아요!
`presentation/ui layer` 주석에 이렇게 적어둔 의도와 data layer 에 둔 채넛의 의도가 궁금합니다.
++ 예시로 permission denied 도 실제로 권한을 사용할 때 activity.permission 어쩌구 api 로 알고 있는데 activity 에서 사용될 친구는 ui 단에서 끊어야 할 것 같다는 생각이 듭니다.

yrsel replied:
패키지 분리 전 에 설계한 코드만 보여주려고 했던 거여서 추후 알맞은 패키지로 이동 시켰습니다!

m6z1 commented on android/app/src/main/java/com/on/turip/ui/common/errorview/CustomErrorView.kt:
```diff
@@ -90,45 +72,18 @@ class CustomErrorView
-        fun setupError(errorType: ErrorEvent) {
-            when (errorType) {
-                ErrorEvent.NETWORK_ERROR -> {
+        fun render(error: UiError) {
```
render라는 이름만 봤을 땐 어떤 역할을 하는지 잘 감이 오지 않는 것 같아요!
주석에 달린 말처럼 에러뷰를 보여주는 것이니 `showErrorView` 이런 네이밍은 어떨지 여쭙습니당

yrsel replied:
좋습니다!
반영 커밋 : https://github.com/woowacourse-teams/2025-Turip/pull/508/commits/69a9657a751da9ea07fd8c25a681c640bff24678

m6z1 reviewed (commented):
꼼꼼한 채넛 고생하셨어요!
노가다성이 많았을 것 같은데 대단해요

제 의견과 궁금한 점을 남겨두었습니다. 

추가로 의견 남겨주셨던 
> 로깅은 api 호출하면 catch 할 때 처리하는 게 좋을 것 같아서, 기존 TuripCustomResult 내부에서 제거하고 아직은 로깅은 추가하지 않았습니다. SafeApiCall 파일의 catch 블럭에서 하는건 어떤가 의견드려요. ( + 로거에 대한 수정 작업도 필요해보이는데 어떤지 궁금합니다. )

이 내용에 대해서 catch 에서 로그 찍는 것 좋은 것 같고 value 값까지 같이 출력하게 하면 될 것 같다는 생각이 들었습니다

yrsel committed:
refactor: ErrorEvent 제거

yrsel committed:
feat: TripDetailActivity 의 state, effect 추가

yrsel committed:
refactor: TripDetailActivity xml 프로그레스바 추가

yrsel committed:
refactor: TripMapper 사용안하는 메서드 제거

yrsel committed:
refactor: TripDetail 화면에 보여주기 위한 UiModel 필드 추가

yrsel committed:
refactor: TripDetailActivity 새로운 에러타입으로 리팩터링

yrsel committed:
refactor: Retrofit 외 API 에 새로운 타입 적용 오류 수정

yrsel committed:
refactor: FavoriteContentFragment 프로그레스바 추가

yrsel committed:
refactor: RegionResultActivity 데이터 존재하는 상황 외에도 지역명은 보이도록 수정

yrsel committed:
refactor: TripDetailActivity 텍스트 뷰 데이터 읽기 전에 접근하지 않도록 수정

yrsel committed:
refactor: HomeScreen 로딩 추가 및 상태 기반으로 화면 분기하며 flow 로 전환

yrsel committed:
refactor: FavoritePlace 화면 상태 데이터 모델링

yrsel committed:
refactor: FavoritePlaceFolderCatalogFragment state 파일 분리, flow로 데이터 옵저빙하도록 수정

yrsel committed:
refactor: SettingScreen 이벤트 처리 리팩터링

yrsel committed:
refactor: FavoritePlaceFolderFragment 에러 상황 스낵바로 알려주도록 수정 및 flow 전환

yrsel committed:
refactor: 폴더 관련 화면들 새로운 에러 타입 적용

yrsel committed:
refactor: 들여쓰기 및 문맥에 안맞는 코드 수정

yrsel committed:
refactor: 폴더 바텀 시트, dismiss effect로 사라지도록 수정

yrsel committed:
refactor: 분산된 UiEffect 파일별로 통합

yrsel committed:
refactor: Result 관련 패키지 수정

yrsel committed:
refactor: renderXXX -> showXXX 함수명 수정

yrsel committed:
refactor: 영상 상세 로딩, 에러 일 때 앱바 보이도록 수정

yrsel committed:
refactor: ktlintformat

yrsel committed:
Merge branch 'feature/#507' of https://github.com/woowacourse-teams/2025-Turip into feature/#507

yrsel committed:
refactor: 프래그먼트 생명주기 flow 확장함수 추가

yrsel committed:
refactor: 에러 상황에서 스낵바 재시도 버튼 항상 제공하도록 수정

yrsel committed:
refactor: 오류 수정

- activity_search.xml : visibility 설정 기본값 gone으로 수정
- FavoriteContentViewModel : 에러타입 오기입 수정
- HomeScreen : 로딩 Modifier 수정
- GoogleCredentialManager : IOException import 수정

yrsel committed:
refactor: TuripResultExtensions 패키지 수

yrsel committed:
refactor: 에러 상황 시, UI에서 행동을 실행하는 게 아닌 ViewModel 내에서 결정하도록 수정

yrsel committed:
refactor: 오류 수정

- TripDetailActivity : 제목이 아예 없는 발생할 수 없는 경우라도 제목 확장 토글 visible 판단 가능하도록 수정
- FavoriteContentViewModel, SettingViewModel, TripDetailViewModel : Flow 내부에서 상태 접근 시 람다 내부에서 접근

yrsel committed:
refactor: ktlintformat

yrsel committed:
refactor: kotlin 스럽게 수정

yrsel reviewed (commented):
PR 본문에 로딩에 대한 의견을 듣고 싶어서 남겼습니다 한 번 읽어주시면 감사하겠습니다
수정내역이 많아서 힘드시겠지만 QA 만이라도 부탁드립니다, 😭 감사합니다!

m6z1 commented on android/app/src/main/java/com/on/turip/ui/common/error/UiError.kt:
```diff
@@ -0,0 +1,50 @@
+         * 서버 또는 클라이언트가 정보 전달하는 과정에서 문제 발생
+         * 서버에 문제가 있음을 알리는 정보를 전달
+         */
+        data object Server : Global
```
단순 질문입니다!
Server 의 주석 설명을 봤을 때 server 가 문제 생겼다는 것은 feature 기능이 작동하며 알 수 있을 것 같은데 global 이란 단위에 포함 시킨 이유가 궁금해요!

yrsel replied:
Global과 Feature를 나눈 기준은 전역에서 발생할 수 있느냐, 아니면 일부 기능에서만 발생할 수 있느냐로 구분했어요.
그렇기에 서버 또는 클라이언트(안드로이드)에서의 문제는 전역에서 발생할 수 있다고 판단해 Global에 포함시켰습니다!

m6z1 commented on android/app/src/main/java/com/on/turip/ui/compose/designsystem/component/ErrorScreen.kt:
```diff
@@ -20,40 +20,18 @@ import androidx.compose.ui.res.stringResource
-                )
-            }
-        }
+    val errorUiModel: ErrorUiModel = errorUiState.toUiModel() ?: return
```
기존 xml 구조를 생각해보면, state 에서 model 로 매핑해주는 역할도 viewModel 에서 진행했었는데, 컴포즈에서는 스크린에서 매핑하도록 만든 이유가 궁금해요!
screen 에 viewModel 을 넘겨 사용하게 할 수도 있지 않나? 싶은 생각이 드는데 채넛의 의도가 궁금합니닷

yrsel replied:
> 기존 xml 구조를 생각해보면, state 에서 model 로 매핑해주는 역할도 viewModel 에서 진행했었는데, 컴포즈에서는 스크린에서 매핑하도록 만든 이유가 궁금해요!

의견을 잘 이해하지 못해서 이해한 바로 답변 남기겠습니다!

1. `기존 xml 구조 = ErrorEvent 사용할 때` 라고 이해한다면,
이 경우에는 onFailure에서 ui에서 사용할 에러타입을 반환해줬기에 model로 매핑할 필요가 없어서 ErrorScreen 자체에서 분기를 통해 리소스들을 설정하는 구조였던 것 같습니다!

2. `ViewModel에서 uiModel을 반환해도 되지 않나?` 라고 이해한다면,
저도 변환이 필요하다면 ViewModel에서 uiModel을 반환하는 게 적절하다고 생각하고 있습니다.
현재 uiState 내부에 화면의 에러 상태(네트워크 오류, 서버 오류,  정상)를 ErrorUiState로 관리하여 표현하고 있는 화면이 존재합니다.
그래서 Screen 또는 Activity 에서 변환해야 하는 상황이 발생하게 되는데 ViewModel, Activity, Screen 여러 군데로 변환하는 지점이 나뉘게 되는 것보다는 ErrorScreen(compose), CustomErrorView(xml) 내부에 있는 게 적절하다고 생각했어요. 
ErrorUiState를 만들지 않고 serverError, networkError에 대한 boolean 으로 관리할 수도 있을 것 같은데, 화면 상태 분기가 너무 복잡해질 것 같아서 ErrorUiState로 묶게 되었습니다.

3. `컴포즈에서는 스크린에서 매핑하도록 만든 이유` 
현재 CustomErrorView, ErrorScreen 모두 동일하게 내부에서 uiModel로 변환하고 있어서 xml과 compose 사용할 때 어떤 부분의 차이를 말씀하신 건 지 이해하지 못했어요..🥲

---

> screen 에 viewModel 을 넘겨 사용하게 할 수도 있지 않나? 싶은 생각이 드는데 채넛의 의도가 궁금합니닷

저는 Screen에서 ViewModel 전부를 알기보다는 필요한 상태만 아는 것이 적절하다고 생각이 들긴 합니다..!

m6z1 replied:
좋아요 errorScreen 에서는 공감합니다. 하지만, 추후에 뷰에 대한 screen 이 존재한다면 그 땐 필요한 애들을 하나 씩 넘겨준다기 보단 viewModel 을 넘기면 필요할 때마다 파라미터를 추가하지 않아도 될 것 같습니다.
SAA 구조로 바뀐다면 route 라는 애가 viewModel 을 가지게 될 테구요!

m6z1 commented on android/app/src/main/java/com/on/turip/ui/compose/home/HomeScreen.kt:
```diff
@@ -150,15 +193,73 @@ fun HomeScreen(
-@Preview(showBackground = true)
+@Preview(showBackground = true, name = "로딩")
+@Composable
+fun Home_Loading_Preview() {
```
따로 preview 네이밍 컨벤션에 대해서 정하지 않았지만 다른 파일에는 파스칼케이스를 여기서는 스네이크 케이스를 써서 통일하면 좋을 것 같아요!

yrsel replied:
반영 커밋 : 023c6e7341dd61aa3bd329dac227b7e805292c46

m6z1 commented on android/app/src/main/java/com/on/turip/ui/compose/home/HomeUiEffect.kt:
```diff
@@ -0,0 +1,5 @@
+package com.on.turip.ui.compose.home
+
+sealed interface HomeUiEffect {
+    data object NavigateToLogin : HomeUiEffect
```
NavigateToLogin 은 모든 뷰에서 넘어갈 수 있을 것 공통적으로 사용될 수 있게 만들어 볼 수 있을까요 ?
예를 들어 계층을 분리할 때 commonUiEffect 를 두고 그 하위에 뷰들의 effect 를 둔다던가.. ?
별로면 패스하셔도 됩니닷

yrsel replied:
원래 commonUiEffect 를 두고 관리되고 있었는데, uiEffect, commonUiEffect 두 개로 나눠서 관리하는 것의 장점을 잘 모르겠어서 화면 단위별 하나의 uiEffect로 관리하도록 수정했어요! 
 
> 계층을 분리할 때 commonUiEffect 를 두고 그 하위에 뷰들의 effect 를 둔다던가

uiEffect에  commonUiEffect를 미리 정의해놓는다는 것 같아 보이는 데 맞나요,,?!
구현하게 된다면 어떤 방식으로 BaseUiEffect(?)가 설계되려나요??

m6z1 replied:
``` kotlin
sealed interface HomeUiEffect {
    data object SpecificAction : HomeUiEffect

    data class Common(val effect: BaseUiEffect) : HomeUiEffect 
}

sealed interface BaseUiEffect {
    data object NavigateToLogin : BaseUiEffect
}
```
요런 느낌 . . . ?

yrsel replied:
Common에 해당하는 effect들이 많이 있으면 유용할 것 같네요!
현재 저희는 NavigateToLogin 하나뿐이고, 그 외 추가될 공통 로직이 떠오르지 않아서 나중에 공통으로 처리할 effect들이 발생할 경우 뭉치의 설계를 반영 해볼게요! 👍

m6z1 commented on android/app/src/main/java/com/on/turip/ui/compose/setting/SettingViewModel.kt:
함수는 동사로 시작하는 걸로..!

yrsel replied:
반영 커밋 : 5bea52136b6ee72cfdda59d61cd5dc7963a82fad

m6z1 reviewed (commented):
채넛 너무너무 고생하셨어요 코드 줄 수만 봐도 ㅠ..,ㅜ정성이 느껴지는 ,,,,

먼저 함수들이 onXXX 로 시작하는 게 있는 것 같아요! 리스너 람다를 제외하곤 동사로 시작하도록 수정 부탁드립니닷

그리고 로딩이 짧아서 깜빡거리는 것처럼 느껴진다고 한 점은 저도 delay 거는 게 좋을 것 같다는 생각이 들었어요 따로 로딩이 짧아서 처리를 해본 적이 없는데 응답이 빠르니 이런 것도 고려를 해야 하네요...😅

기기로 돌려보는데 게스트 모드는 되지만 구글 로그인이 안 되어요 .. ㅠ.,ㅜ 계정 선택해도 홈 뷰로 넘어가지 않습니다... 요 부분 확인 한 번 해주실 수 있으실까요 ??

yrsel committed:
refactor: LoginScreen event -> effect 로 네이밍 통일

yrsel committed:
refactor: HomePreview 파스칼 케이스로 네이밍 수정

yrsel committed:
refactor: Screen Preview를 위한 컴포넌트 분리 제거

- Preview 내부에서 theme 또는 Scaffold 감싸서 확인하도록 수정

yrsel committed:
refactor: 함수명 동사로 수정

yrsel committed:
refactor: Failure Throwable nullable 하지 않도록 수정

yrsel committed:
refactor: ktlintformat

yrsel committed:
refactor: FavoritePlaceFolderCatalogViewModel 기본값이 아닌 직접 명시해주도록 수정

yrsel committed:
refactor: HomeScreen Insets 조절

- Fragment 위에 ComposeView를 띄우는 구조여서 Insets는 Activity, Fragment에서 적용되고, Compose에서는 기본값으로 설정되도록 수정

m6z1 reviewed (approved):
믿음의 LGTM

yrsel committed:
refactor: 홈화면 지역 선택시, 화면 전체에 로딩이 반영되는 부분 제거

yrsel committed:
refactor: 로딩 -> 정상 상태 변경시, 깜빡이는 문제 해결

yrsel committed:
refactor: ktlintformat

yrsel committed:
refactor: 찜 폴더 장소 비어있을 때 분기 처리 위치 수정, 현재 폴더 재클릭 시, API 호출 안하도록 수정

yrsel committed:
refactor: HomeScreen 중복 정의된 패딩 값 제거

yrsel committed:
refactor: empty 화면 상태 분기 추가

yrsel committed:
refactor: 정상 API 호출 완료시에만 에러 상태 변경하도록 수정

yrsel commented on android/app/src/main/java/com/on/turip/ui/main/favorite/FavoriteContentFragment.kt:
```diff
@@ -78,69 +78,81 @@ class FavoriteContentFragment : BaseFragment<FragmentFavoriteContentBinding>() {
-            handleVisibleByHasContent(favoriteContents)
-            favoriteContentAdapter.submitList(favoriteContents)
+        collectOnStarted(viewModel.uiState) { uiState: FavoriteContentUiState ->
+            if (uiState.isLoading) showLoading()
+            when {
+                uiState.errorUiState != ErrorUiState.None -> showErrorView(uiState.errorUiState)
+                uiState.isEmpty -> showEmptyView()
+                else -> showContents(uiState.favoriteContents)
+            }
```
when 문 분기로 인해서 Loading 상태로 잠시 존재하다가 API가 순식간에 성공해 Success로 되는 경우 반드시 깜빡이게 되어서 바깥으로 뺐어요
로딩은 성공, 에러, 빈 화면과는 공존할 수 있다고 생각해요. 
공존하게 된다면 로딩이 성공, 에러, 빈 화면을 위에 존재하다가 로딩이 사라지는 방식이 될 것 같은데, 
화면 상태 관리를 위해 어떤 방식으로 접근하는 게 좋을까요 ?

m6z1 commented on android/app/src/main/java/com/on/turip/ui/compose/home/HomeScreen.kt:
```diff
@@ -96,6 +96,14 @@ private fun HomeScreenContent(
     val focusManager = LocalFocusManager.current
     val keyboardController = LocalSoftwareKeyboardController.current
 
+    if (uiState.isLoading) {
```
👍 👍 👍 최고다 채넛 💯

m6z1 reviewed (approved):
💯 💯 💯 💯

yrsel committed:
refactor: 로그인 시, 네트워크 없으면 스낵바로 알리기

yrsel committed:
refactor: 최초 상태가 아닌 비어 있는 상태를 판별하도록 수정

yrsel committed:
refactor: ci 오류 해결
